File: app/src/main/java/com/vereshchagin/nikolay/stankinschedule/ui/schedule/model/pair/DateItem.java
package com.vereshchagin.nikolay.stankinschedule.ui.schedule.model.pair;

import android.os.Parcelable;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.gson.JsonArray;

import java.util.Calendar;

/**
 * Абстактный класс дат пары в расписании.
 */
public abstract class DateItem implements Comparable<DateItem>, Parcelable {

    static final String JSON_DATE = "date";
    static final String JSON_FREQUENCY = "frequency";

    /**
     * @return день недели даты.
     */
    @NonNull
    public abstract DayOfWeek dayOfWeek();

    /**
     * @return возвращает полную дату в формате: YYYY.MM.DD.
     */
    @NonNull
    public abstract String fullDate();

    /**
     * @return первый день даты.
     */
    @NonNull
    public abstract Calendar firstDay();

    /**
     * @return последний день даты.
     */
    @NonNull
    public abstract Calendar lastDay();

    /**
     * @return периодичность даты.
     */
    @NonNull
    public abstract FrequencyEnum frequency();

    /**
     * Определяет, пересекаются ли даты между собой.
     * @param dateItem сравниваемая дата.
     * @return true если пересекаются, иначе false.
     */
    public abstract boolean intersect(@NonNull DateItem dateItem);

    /**
     * Добавляет DateItem в json массив.
     * @param array json массив.
     */
    public abstract void toJson(@NonNull JsonArray array);

    @Override
    public abstract boolean equals(@Nullable Object o);
}

File: app/src/main/java/com/vereshchagin/nikolay/stankinschedule/model/schedule/pair/DateItem.kt
package com.vereshchagin.nikolay.stankinschedule.model.schedule.pair

import android.os.Parcelable
import com.google.gson.JsonObject
import com.vereshchagin.nikolay.stankinschedule.model.schedule.json.JsonPairItem

/**
 * Абстрактный класс дат пары в расписании.
 */
sealed class DateItem : Comparable<DateItem>, Parcelable {

    internal companion object {
        const val JSON_DATE = "date"
        const val JSON_FREQUENCY = "frequency"
        const val JSON_DATE_PATTERN = "yyyy.MM.dd"
        const val JSON_DATE_PATTERN_V2 = "yyyy-MM-dd"
    }

    /**
     * Возвращает день недели даты.
     */
    abstract fun dayOfWeek(): DayOfWeek

    /**
     * Возвращает периодичность даты.
     */
    abstract fun frequency(): Frequency

    /**
     * Возвращает JsonObject дату.
     */
    abstract fun toJson(): JsonObject

    /**
     * Возвращает Json дату.
     */
    abstract fun toJsonItem(): JsonPairItem.JsonDateItem

    /**
     * Определяет, пересекаются ли даты между собой.
     */
    abstract fun intersect(item: DateItem): Boolean

    /**
     * Определяет, находиться ли дата раньше другой.
     */
    abstract fun isBefore(item: DateItem): Boolean

    /**
     * Возвращает копию объекта.
     */
    abstract fun clone() : DateItem

    override fun compareTo(other: DateItem): Int {
        if (this == other) {
            return 0
        }
        return if (isBefore(other)) -1 else 1
    }

    abstract override fun equals(other: Any?): Boolean

    abstract override fun hashCode(): Int

    abstract override fun toString(): String
}